src/App.tsx

```diff
@@ -1,26 +1,293 @@
+    } finally {
+      setLoadingTodoIds(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  const onAddTodo = async (todoTitle: string) => {
+    setTempTodo({ id: 0, title: todoTitle, completed: false, userId: USER_ID });
+    try {
+      const newTodo = await addTodos({ title: todoTitle, completed: false });
+
+      setTodos(item => [...item, newTodo]);
+    } catch (err) {
+      setError(Error.Add);
+      throw err;
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const onClearCompleted = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => {
+      onDeleteTodo(todo.id);
+    });
+  };
+
+  const onUpdateTodo = async (todoUpdate: Todo) => {
+    const oldTodos = [...todos];
+
+    setLoadingTodoIds(prev => [...prev, todoUpdate.id]);
+
+    try {
+      await updateTodo(todoUpdate);
+      setTodos(currentTodos =>
+        currentTodos.map(todo => {
+          return todo.id === todoUpdate.id ? todoUpdate : todo;
+        }),
+      );
+    } catch (err) {
+      setTodos(oldTodos);
+      setError(Error.Update);
+      throw err;
+    } finally {
+      setLoadingTodoIds(prev => prev.filter(id => id !== todoUpdate.id));
+    }
+  };
+
+  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (inputValue.trim() === '') {
+      setError(Error.Empty);
+
+      return;
+    }
+
+    try {
+      await onAddTodo(inputValue.trim());
+      setInputValue('');
+    } catch (err) {}
+  };
+
+  const inputFocus = () => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  };
+
+  useEffect(() => {
+    if (todos && !editingTodoId) {
+      inputFocus();
+    }
+  }, [error, todos, editingTodoId]);
+
+  useEffect(() => {
+    const timer = setTimeout(() => setError(Error.Reset), 3000);
+
+    return () => clearTimeout(timer);
+  }, [error]);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setError(Error.Load));
+  }, []);
+
+  const filteredTodos = todos.filter(todo => {
+    if (filter === Filter.Active) {
+      return !todo.completed;
+    }
+
+    if (filter === Filter.Completed) {
+      return todo.completed;
+    }
+
+    return Filter.All;
+  });
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const unCompletedTodosCounter = todos.filter(todo => !todo.completed).length;
+  const todosCompletedNum = todos.filter(todo => todo.completed).length;
+  const allTodosCompleted =
+    todos.length > 0 && todos.every(todo => todo.completed);
+
+  const onToggleAll = () => {
+    if (unCompletedTodosCounter > 0) {
+      const activeTodos = todos.filter(todo => !todo.completed);
+
+      activeTodos.forEach(todo => {
+        onUpdateTodo({ ...todo, completed: true });
+      });
+      setIsToogleAll(true);
+    } else {
+      todos.forEach(todo => {
+        onUpdateTodo({ ...todo, completed: false });
+      });
+      setIsToogleAll(false);
```

Comment:
seems very similar can let's write one time and just change true or false depends on unCompletedTodos counter